numba/targets/imputils.py

```diff
@@ -72,6 +72,23 @@ def imp(context, builder, sig, args):
             builder, func, fndesc.restype, fndesc.argtypes, args, env=None)
         with cgutils.if_unlikely(builder, status.is_error):
             context.call_conv.return_status_propagate(builder, status)
+        assert sig.return_type == fndesc.restype
+        # Reconstruct optional return type
+        if isinstance(sig.return_type, types.Optional):
```

Comment:
Can you explain why this is needed, while fndesc.restype is already an Optional?